### DISCUSSION TITLE
history module privacy issue / metadata black hole

### DISCUSSION THREAD

**Discussion by elandorr** [rspamd/rspamd#4480] [Q&A] [answered]
The history module has a row limit of 200 by default.

I manually looked into the database, decompressed it, and found data back a long time ago.

**That's an exorbitant privacy problem.** The docs mention nothing the like: https://www.rspamd.com/doc/modules/history_redis.html .
Tasty metadata going back very far, all without anyone expecting it to after reading the docs.

Clearly the 200 only refer to the webui.

1. When is the history actually deleted?

It looks like it goes on forever. 

2. Can you have the history module enabled, and it automatically delete older than x?

3. How does the database work in the first place? The records are in random order and compressed - how does rspamd turn that into sorted output? I checked the lua code but couldn't make sense of it.

4. As the redis isn't sortable, how does this not destroy performance on bigger instances, if it goes on indefinitely?

Thanks in advance.

`rs_historyHOSTNAME_zst` contains the last 200 rows, in order, as expected

`rs_historyRANDOMSTRING_zst` contains varying amounts of emails, and apparently an  unlimited amount of itself

I cannot find the code responsible for this, it's driving me nuts. Can someone enlighten me?

**Comment by vstakhov** [comment]
> `rs_historyRANDOMSTRING_zst` contains varying amounts of emails, and apparently an unlimited amount of itself
> 
> I cannot find the code responsible for this, it's driving me nuts. Can someone enlighten me?

What is `rs_historyRANDOMSTRING_zst`? It might be a leftover from some old versions of Rspamd...

**Comment by elandorr** [comment]
@vstakhov Thanks for your reply!

I don't know if that could be a hash that could leak anything. It's 12 random alphanumeric chars.
Here's one with random letters switched:
`rs_historyab2x20xc69x3_zst` It's a list type record. This list then has varying amounts of e-mail records, in no particular order as far as I can tell.

This string is not greppable in the mails.

> It might be a leftover from some old versions of Rspamd...

I thought so too, but even the very latest e-mails are included in it.

I checked the source up and down to find where this entry could be made, but I can't find it.

**Comment by elandorr** [comment]
If there's anything I can do to help figure this out, besides handing over userdata which is obviously not possible, let me know.

New mails continue to be dumped in these mystery records and we have to stop it.

**Comment by elandorr** [comment]
I realized something new: The mysterious records contain at most 200 entries. It seems they are a kind of snapshot of the expected `rs_historyHOSTNAME_zst`. Still no clue where they come from.

Not sure I should ping anyone, but we ought to get rid of this metadata quickly. It goes back many thousands of mails.

**Comment by elandorr** [comment, marked as answer]
Summary: It was an accident in the used stack.
1. rspamd works fine
3. the limit is respected
4. It's trivial and just pushes/trims
5. not relevant anymore

I'm pretty sure I figured it out: The stack we use previously used a randomly generated hostname. Every update/reboot triggered a new record that looks like a snapshot. Not sure why that was ever saved, I think it wasn't meant to be persistent at all so I'd guess mem-only, but it might just be a human mistake.

If I don't just dump all the contents, but also dump the key names, the fresh mails seem to be in the expected place and respect the 200 limit.

Goes to show what slips through the cracks in changelogs - I'll report this in the right place. I doubt many people audit the db like me, so there are probably thousands upon thousands of tasty metadata stashes.

Have a nice day guys.
